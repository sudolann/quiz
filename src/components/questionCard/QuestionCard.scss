@import '../../styles/constants.scss';

.question {
  display: flex;
  width: 100%;
  height: 100%;
  padding: 3rem 0;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  min-height: 300px;
  box-sizing: content-box;

  @media (min-width: $breakpoint-tablet) {
    min-height: 350px;
  }

  @media (min-width: $breakpoint-desktop) {
    min-height: 400px;
  }
}

.question__card {
  box-shadow: 0px 0px 8px 0px rgba(130, 130, 130, 0.64);
  border-radius: 0.35rem;
  width: 20rem;
  height: 12rem;
  position: relative;
  box-sizing: content-box;

  @media (min-width: $breakpoint-tablet) {
    width: 24rem;
    height: 14.5rem;
  }

  @media (min-width: $breakpoint-desktop) {
    width: 31.2rem;
    height: 18.5rem;
  }

  p {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -80%);
    width: 70%;
    text-align: center;
  }
}

.question__card__buttons {
  border-top: 1px solid rgba(130, 130, 130, 0.19);
  height: 2.8rem;
  width: 100%;
  position: absolute;
  bottom: 0;
  left: 0;
  display: flex;
  justify-content: space-around;
  align-items: center;

  @media (min-width: $breakpoint-tablet) {
    height: 4rem;
  }

  @media (min-width: $breakpoint-desktop) {
    height: 4.9rem;
  }
}

.question__footer p {
  font-size: 0.85rem;

  @media (min-width: $breakpoint-tablet) {
    font-size: 1rem;
  }

  @media (min-width: $breakpoint-desktop) {
    font-size: 1.3rem;
  }
}
