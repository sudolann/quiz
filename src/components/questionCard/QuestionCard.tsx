import React, { FunctionComponent, ReactElement } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button } from 'antd';
import { getQuizData } from '../../state/quiz/selectors';
import { QuizState } from '../../state/quiz/reducer';
import { setScore } from '../../state/points/actions';
import { getPoints } from '../../state/points/selectors';
import { Redirect } from 'react-router';
import './QuestionCard.scss';

export const QuestionCard: FunctionComponent = (): ReactElement => {
  const dispatch = useDispatch();
  const quiz: QuizState = useSelector(getQuizData);
  const points: Array<0 | 1> = useSelector(getPoints);
  const list = quiz.list;
  const handleClick = (answer: boolean) => {
    const correctAnswer = list[points.length].correct_answer === 'True' ? true : false;
    const point = correctAnswer === answer ? 1 : 0;
    dispatch(setScore(point));
  };

  if (list.length === points.length) {
    return <Redirect to='/results' />;
  }
  return (
    <div className='question'>
      <div className='question__header'>
        <h1>{list[points.length]?.category}</h1>
      </div>
      <div className='question__card'>
        <p>{list[points.length]?.question}</p>
        <div className='question__card__buttons'>
          <Button type='primary' ghost id='quiz-btn' onClick={(): void => handleClick(true)}>
            true
          </Button>
          <Button type='primary' ghost id='quiz-btn' onClick={(): void => handleClick(false)}>
            false
          </Button>
        </div>
      </div>
      <div className='question__footer'>
        <p>
          {points.length + 1} of {list.length}
        </p>
      </div>
    </div>
  );
};
