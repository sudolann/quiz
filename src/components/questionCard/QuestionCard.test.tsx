import { shallow } from 'enzyme';
import React from 'react';
import { QuestionCard } from './QuestionCard';
import renderer from 'react-test-renderer';
import * as redux from 'react-redux';
import { Redirect } from 'react-router';

const mockDispatch = jest.fn();
const spy = jest.spyOn(redux, 'useSelector');
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: (): jest.Mock<any, any> => mockDispatch,
}));
const quiz = {
  list: [
    {
      category: 'General Knowledge',
      question: 'Nutella is produced by the German company Ferrero.',
      correct_answer: 'False',
    },
    { category: 'General Knowledge', question: 'Romanian belongs to the Romance language family, shared with French, Spanish, Portuguese and Italian. ', correct_answer: 'True' },
  ],
  loading: false,
};

describe('component', (): void => {
  describe('QuestionCard', (): void => {
    it('match snapshot', (): void => {
      spy.mockReturnValue({
        quiz,
        points: [],
      });
      const tree = renderer.create(<QuestionCard />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders header', (): void => {
      spy.mockReturnValue({
        quiz,
        points: [],
      });
      const wrapper = shallow(<QuestionCard />);
      const h1 = wrapper.find('h1');
      expect(h1.text()).toBe(quiz.list[0].category);
    });
    it('renders Question', (): void => {
      spy.mockReturnValue({
        quiz,
        points: [],
      });
      const wrapper = shallow(<QuestionCard />);
      const h1 = wrapper.find('.question__card p');
      expect(h1.text()).toBe(quiz.list[0].question);
    });
    it('renders footer', (): void => {
      const wrapper = shallow(<QuestionCard />);
      const h1 = wrapper.find('.question__footer p');
      expect(h1.text()).toBe('1 of 2');
    });
    it('call dispatch when button was clicked', (): void => {
      spy.mockReturnValue({
        quiz,
        points: [],
      });
      const wrapper = shallow(<QuestionCard />);
      wrapper
        .find('Button[id="quiz-btn"]')
        .at(1)
        .simulate('click');
      wrapper
        .find('Button[id="quiz-btn"]')
        .at(0)
        .simulate('click');
      expect(mockDispatch).toBeCalledTimes(2);
    });
    it('redirect to results component when there are two sores in the state', (): void => {
      spy.mockReturnValue({
        quiz,
        points: [0, 1],
      });
      const wrapper = shallow(<QuestionCard />);
      expect(wrapper.containsMatchingElement(<Redirect to='/results' />));
    });
  });
});
