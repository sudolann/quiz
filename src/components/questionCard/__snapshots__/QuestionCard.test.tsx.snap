// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component QuestionCard match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="question"
  >
    <div
      className="question__header"
    >
      <h1>
        General Knowledge
      </h1>
    </div>
    <div
      className="question__card"
    >
      <p>
        Nutella is produced by the German company Ferrero.
      </p>
      <div
        className="question__card__buttons"
      >
        <button
          className="ant-btn ant-btn-primary ant-btn-background-ghost"
          id="quiz-btn"
          onClick={[Function]}
          type="button"
        >
          <span>
            true
          </span>
        </button>
        <button
          className="ant-btn ant-btn-primary ant-btn-background-ghost"
          id="quiz-btn"
          onClick={[Function]}
          type="button"
        >
          <span>
            false
          </span>
        </button>
      </div>
    </div>
    <div
      className="question__footer"
    >
      <p>
        1
         of 
        2
      </p>
    </div>
  </div>
</div>
`;
